<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.12"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Genom-1: Sequence Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">Genom-1
   </div>
   <div id="projectbrief">We were given the objective to build a program with an application in actual genomic problems. Several inputs can be given by the user and the idea is to offer different outputs that can help the user to solve his problem. The output can be a list of sites, affinity score for a given sequence, a PWM (or a PSSM), a graphic representation of the PMW (logo), etc.</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.12 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="class_sequence-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">Sequence Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p><code>#include &lt;<a class="el" href="sequence_8hpp_source.html">sequence.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a532b7e8df6ff6b2f990c14ae97859ca2"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_sequence.html#a532b7e8df6ff6b2f990c14ae97859ca2">Sequence</a> ()</td></tr>
<tr class="separator:a532b7e8df6ff6b2f990c14ae97859ca2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aee09a7d70c3ab523fed85da94ea1366f"><td class="memItemLeft" align="right" valign="top">virtual&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_sequence.html#aee09a7d70c3ab523fed85da94ea1366f">~Sequence</a> ()</td></tr>
<tr class="separator:aee09a7d70c3ab523fed85da94ea1366f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9fa615dfb02cbf117604a91dcf9bdd9c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_sequence.html#a9fa615dfb02cbf117604a91dcf9bdd9c">outputSites</a> (const vector&lt; <a class="el" href="struct_pos_dir.html">PosDir</a> &gt; &amp;) const</td></tr>
<tr class="separator:a9fa615dfb02cbf117604a91dcf9bdd9c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa7d40f02696c8e8d25c4fb56220e8ef1"><td class="memItemLeft" align="right" valign="top">vector&lt; <a class="el" href="struct_pos_dir.html">PosDir</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_sequence.html#aa7d40f02696c8e8d25c4fb56220e8ef1">motifRecognition</a> (const string &amp;, const string &amp;fileName) const</td></tr>
<tr class="separator:aa7d40f02696c8e8d25c4fb56220e8ef1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a686c72210c48cc00f8a233e051ec3580"><td class="memItemLeft" align="right" valign="top">vector&lt; <a class="el" href="struct_pos_dir.html">PosDir</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_sequence.html#a686c72210c48cc00f8a233e051ec3580">motifRecognition</a> (const vector&lt; <a class="el" href="struct_coordinate.html">Coordinate</a> &gt; &amp;coordinates, const string &amp;fileName) const</td></tr>
<tr class="separator:a686c72210c48cc00f8a233e051ec3580"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a24d22fd157aaecab26f13342b03fe93e"><td class="memItemLeft" align="right" valign="top">vector&lt; char &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_sequence.html#a24d22fd157aaecab26f13342b03fe93e">giveReverseComplementarySeq</a> (const vector&lt; char &gt; &amp;) const</td></tr>
<tr class="separator:a24d22fd157aaecab26f13342b03fe93e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a308c6c7902f7f587a2d7b2d01643772f"><td class="memItemLeft" align="right" valign="top">vector&lt; <a class="el" href="struct_coordinate.html">Coordinate</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_sequence.html#a308c6c7902f7f587a2d7b2d01643772f">readBedGraph</a> (const string &amp;fileName)</td></tr>
<tr class="separator:a308c6c7902f7f587a2d7b2d01643772f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6fb228a42c420eeede32a35925bb7b16"><td class="memItemLeft" align="right" valign="top">vector&lt; <a class="el" href="struct_bed_coordinate.html">BedCoordinate</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_sequence.html#a6fb228a42c420eeede32a35925bb7b16">ReadBed</a> (const string &amp;fileName)</td></tr>
<tr class="separator:a6fb228a42c420eeede32a35925bb7b16"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8e1e675d608ecaf6a7b99e3def92d1ad"><td class="memItemLeft" align="right" valign="top">vector&lt; string &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_sequence.html#a8e1e675d608ecaf6a7b99e3def92d1ad">scanFasta</a> (vector&lt; <a class="el" href="struct_coordinate.html">Coordinate</a> &gt; &amp;coordinates, const string &amp;fileName, int nbrOfSeq)</td></tr>
<tr class="separator:a8e1e675d608ecaf6a7b99e3def92d1ad"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad43e0f0222082991a8ef186c3add8007"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_sequence.html#ad43e0f0222082991a8ef186c3add8007">loadResultsOnFile</a> (const string &amp;fileName, <a class="el" href="struct_pos_dir.html">PosDir</a> posdir, double sommeScores)</td></tr>
<tr class="separator:ad43e0f0222082991a8ef186c3add8007"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a416ffd3a8a824144fc7696099bcabd96"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_sequence.html#a416ffd3a8a824144fc7696099bcabd96">loadMatrixOnFile</a> (const string &amp;fileName, <a class="el" href="matrix_8hpp.html#abdbfe8c3065564355f01ac70dd548716">matrix</a> matrice)</td></tr>
<tr class="separator:a416ffd3a8a824144fc7696099bcabd96"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af79afc5598e60f53d7887800a0e8b680"><td class="memItemLeft" align="right" valign="top">vector&lt; <a class="el" href="struct_pos_dir.html">PosDir</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_sequence.html#af79afc5598e60f53d7887800a0e8b680">find</a> (<a class="el" href="class_matrix_protein.html">MatrixProtein</a> &amp;, const string &amp;, const string &amp;, const double &amp;)</td></tr>
<tr class="separator:af79afc5598e60f53d7887800a0e8b680"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="a532b7e8df6ff6b2f990c14ae97859ca2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a532b7e8df6ff6b2f990c14ae97859ca2">&sect;&nbsp;</a></span>Sequence()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">Sequence::Sequence </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="aee09a7d70c3ab523fed85da94ea1366f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aee09a7d70c3ab523fed85da94ea1366f">&sect;&nbsp;</a></span>~Sequence()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">Sequence::~Sequence </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="af79afc5598e60f53d7887800a0e8b680"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af79afc5598e60f53d7887800a0e8b680">&sect;&nbsp;</a></span>find()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt; <a class="el" href="struct_pos_dir.html">PosDir</a> &gt; Sequence::find </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="class_matrix_protein.html">MatrixProtein</a> &amp;&#160;</td>
          <td class="paramname"><em>mtrx</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>fasta</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const double &amp;&#160;</td>
          <td class="paramname"><em>threshold</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>find method take a matrix, loads it, goes through sequences in fasta and gets all motifs with sufficient scores with their positions and direction </p>

</div>
</div>
<a id="a24d22fd157aaecab26f13342b03fe93e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a24d22fd157aaecab26f13342b03fe93e">&sect;&nbsp;</a></span>giveReverseComplementarySeq()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt; char &gt; Sequence::giveReverseComplementarySeq </td>
          <td>(</td>
          <td class="paramtype">const vector&lt; char &gt; &amp;&#160;</td>
          <td class="paramname"><em>seq</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>The giveComplementarySeq method gives the reverse complementary sequence of any nucleotidic sequences, whether a motif or a genomic sequence. </p>

</div>
</div>
<a id="a416ffd3a8a824144fc7696099bcabd96"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a416ffd3a8a824144fc7696099bcabd96">&sect;&nbsp;</a></span>loadMatrixOnFile()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Sequence::loadMatrixOnFile </td>
          <td>(</td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>fileName</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="matrix_8hpp.html#abdbfe8c3065564355f01ac70dd548716">matrix</a>&#160;</td>
          <td class="paramname"><em>matrice</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>loadMatrixOnFile method loads on a given file a given matrix </p>

</div>
</div>
<a id="ad43e0f0222082991a8ef186c3add8007"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad43e0f0222082991a8ef186c3add8007">&sect;&nbsp;</a></span>loadResultsOnFile()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Sequence::loadResultsOnFile </td>
          <td>(</td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>fileName</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_pos_dir.html">PosDir</a>&#160;</td>
          <td class="paramname"><em>posdir</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>sommeScores</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>loadResultsOnFile method loads on a given file all the informations related to a given sequence/several sequences </p>

</div>
</div>
<a id="aa7d40f02696c8e8d25c4fb56220e8ef1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa7d40f02696c8e8d25c4fb56220e8ef1">&sect;&nbsp;</a></span>motifRecognition() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt; <a class="el" href="struct_pos_dir.html">PosDir</a> &gt; Sequence::motifRecognition </td>
          <td>(</td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>motif</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>fileName</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>The motifRecognition method opens a .fasta file and find a given motif on a sequence (and its complement) in the .fasta file. </p>

</div>
</div>
<a id="a686c72210c48cc00f8a233e051ec3580"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a686c72210c48cc00f8a233e051ec3580">&sect;&nbsp;</a></span>motifRecognition() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt;<a class="el" href="struct_pos_dir.html">PosDir</a>&gt; Sequence::motifRecognition </td>
          <td>(</td>
          <td class="paramtype">const vector&lt; <a class="el" href="struct_coordinate.html">Coordinate</a> &gt; &amp;&#160;</td>
          <td class="paramname"><em>coordinates</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>fileName</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>The motifRecognition method can also open a .fasta file and, with a list of possible motifs, find which one are present within the sequences of the .fasta file. </p>

</div>
</div>
<a id="a9fa615dfb02cbf117604a91dcf9bdd9c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9fa615dfb02cbf117604a91dcf9bdd9c">&sect;&nbsp;</a></span>outputSites()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Sequence::outputSites </td>
          <td>(</td>
          <td class="paramtype">const vector&lt; <a class="el" href="struct_pos_dir.html">PosDir</a> &gt; &amp;&#160;</td>
          <td class="paramname"><em>info</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>The outputSite method outputs a file containing sequence numbers (from a .fasta file) and the position on which a motif is found, as well as the direction of the reading frame (+ means forward, - means reverse). </p>

</div>
</div>
<a id="a6fb228a42c420eeede32a35925bb7b16"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6fb228a42c420eeede32a35925bb7b16">&sect;&nbsp;</a></span>ReadBed()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt;<a class="el" href="struct_bed_coordinate.html">BedCoordinate</a>&gt; Sequence::ReadBed </td>
          <td>(</td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>fileName</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>The ReadBed method lets us read a .bed file to find the start and end positions of a motif on a chromosome (without any score). </p>

</div>
</div>
<a id="a308c6c7902f7f587a2d7b2d01643772f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a308c6c7902f7f587a2d7b2d01643772f">&sect;&nbsp;</a></span>readBedGraph()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt; <a class="el" href="struct_coordinate.html">Coordinate</a> &gt; Sequence::readBedGraph </td>
          <td>(</td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>fileName</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>The readBedGraph method lets us read a .bedGraph file to find the start and end positions of a motif on a chromosome, along with a score. </p>

</div>
</div>
<a id="a8e1e675d608ecaf6a7b99e3def92d1ad"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8e1e675d608ecaf6a7b99e3def92d1ad">&sect;&nbsp;</a></span>scanFasta()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt; string &gt; Sequence::scanFasta </td>
          <td>(</td>
          <td class="paramtype">vector&lt; <a class="el" href="struct_coordinate.html">Coordinate</a> &gt; &amp;&#160;</td>
          <td class="paramname"><em>coordinates</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>fileName</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>nbrOfSeq</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>The scanFasta method uses a list of coordinates from a .bedgrah file to scan each chromosome genomic sequence in a .fasta file for regions listed in the .bedgraph. </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>src/<a class="el" href="sequence_8hpp_source.html">sequence.hpp</a></li>
<li>src/<a class="el" href="sequence_8cpp.html">sequence.cpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.12
</small></address>
</body>
</html>
